packages/react-paypal-js/src/v6/components/PayPalProvider.tsx

```diff
@@ -0,0 +1,247 @@
+ *
+ * @example
+ * <PayPalSdkInstanceProvider
+ *   components={["paypal-payments"]}
+ *   clientToken={token}
+ *   scriptOptions={{ environment: "sandbox" }}
+ * >
+ *   {children}
+ * </PayPalSdkInstanceProvider>
+ */
+export const PayPalProvider: React.FC<PayPalProviderProps> = ({
+    clientMetadataId,
+    clientToken,
+    components,
+    locale,
+    pageType,
+    partnerAttributionId,
+    shopperSessionId,
+    testBuyerCountry,
+    children,
+    scriptOptions,
+}) => {
+    // Auto-memoize props based on deep equality to prevent unnecessary reloads
+    const memoizedCreateOptions = useDeepCompareMemoize({
+        clientMetadataId,
+        clientToken,
+        components,
+        locale,
+        pageType,
+        partnerAttributionId,
+        shopperSessionId,
+        testBuyerCountry,
+    });
+    const memoizedScriptOptions = useDeepCompareMemoize(scriptOptions);
+
+    // Track if we've already handled the initial hydration
+    const hasHandledInitialHydration = useRef(false);
+
+    const [state, dispatch] = useReducer(instanceReducer, {
+        sdkInstance: null,
+        eligiblePaymentMethods: null,
+        loadingStatus: INSTANCE_LOADING_STATE.INITIAL,
+        error: null,
+        createInstanceOptions: memoizedCreateOptions,
+        scriptOptions: memoizedScriptOptions,
+    });
+
+    // Client-side hydration: transition from INITIAL to PENDING state
+    useEffect(() => {
+        if (
+            !isServer() &&
+            state.loadingStatus === INSTANCE_LOADING_STATE.INITIAL &&
+            !hasHandledInitialHydration.current
+        ) {
+            hasHandledInitialHydration.current = true;
+            dispatch({
+                type: INSTANCE_DISPATCH_ACTION.SET_LOADING_STATUS,
+                value: INSTANCE_LOADING_STATE.PENDING,
+            });
+        }
+    }, [state.loadingStatus]); // Run when loadingStatus changes, but only act once
+
+    // Auto-sync createInstanceOptions changes (e.g., client token updates)
+    useEffect(() => {
+        const hasOptionsChanged =
+            state.createInstanceOptions !== memoizedCreateOptions ||
+            state.scriptOptions !== memoizedScriptOptions;
+
+        if (hasOptionsChanged) {
+            dispatch({
+                type: INSTANCE_DISPATCH_ACTION.RESET_STATE,
+                value: {
+                    createInstanceOptions: memoizedCreateOptions,
+                    scriptOptions: memoizedScriptOptions,
+                },
+            });
+        }
+    }, [
+        memoizedCreateOptions,
+        memoizedScriptOptions,
+        state.createInstanceOptions,
+        state.scriptOptions,
+    ]);
+
+    // SDK loading effect - only runs on client (useEffect doesn't run during SSR)
+    useEffect(() => {
+        if (
+            state.loadingStatus !== INSTANCE_LOADING_STATE.PENDING ||
+            state.sdkInstance
+        ) {
+            return;
+        }
+
+        let isSubscribed = true;
```

Comment:
Can you explain what this does exactly? To my knowledge once the hook calls the cleanup function this useEffect will be lost so it shouldn't run anymore. Do we need this `isSubscribed` logic?